libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h

```diff
@@ -561,6 +554,71 @@ inline _CUDA_VSTD::uint64_t * barrier_native_handle(barrier<thread_scope_block>
     return reinterpret_cast<_CUDA_VSTD::uint64_t *>(&b.__barrier);
 }
 
+
+template <typename _CF = _CUDA_VSTD::__empty_completion>
+_LIBCUDACXX_NODISCARD_ATTRIBUTE _LIBCUDACXX_DEVICE inline
+typename barrier<thread_scope_block, _CF>::arrival_token arrive_tx(
+    barrier<thread_scope_block, _CF> & __b,
+    _CUDA_VSTD::ptrdiff_t __arrive_count_update,
+    _CUDA_VSTD::ptrdiff_t __transaction_count_update) {
+
+    // For ABI reasons, we add the _CF template parameter, but do not yet
+    // support using it.
+    static_assert(std::is_same<_CF, _CUDA_VSTD::__empty_completion>::value, "arrive_tx does not support custom completion functions.");
```

Comment:
What ABI reason? The current ABI of barriers with non-default completion functions is that they are *not* mbarriers. If we want to change that, we need to bump the ABI version anyway and then we can break this.

Drop the `_CF` parameter and this static assert.

Reply:
If we don't make this a template in the CTK 12.x, can we make an ABI breaking change to make this API a template during the 12.x cycle or do we need to wait for CTK 13.x to make that change?

Reply:
We can break ABI arbitrarily (since we provide ways to go back to an older ABI, within those shipped within the current and previous major CUDA version).

Reply:
Thanks. TBH I find libcu++ ABI evolution docs confusing. They start by saying "no guarantees whatsoever", but then they restrict that to "approximately one year", "every major release", and then "but long term support for an ABI for two years", and finally they point to a dead link to read more about it..